crates/ruff_dev/src/format_project.rs

```diff
@@ -0,0 +1,195 @@
+//! This script takes an project and formats all relevant python files in it. What counts as a
+//! relevant pythons file are determined by ruff's default linter settings
+
+#![allow(clippy::print_stdout)]
+
+use anyhow::{bail, Context};
+use clap::Parser;
+use indicatif::ProgressBar;
+use log::debug;
+use rayon::iter::IntoParallelIterator;
+use rayon::iter::ParallelIterator;
+use ruff::resolver::python_files_in_path;
+use ruff::settings::types::{FilePattern, FilePatternSet};
+use ruff_cli::args::CheckArgs;
+use ruff_cli::resolve::resolve;
+use ruff_python_formatter::{format_module, PyFormatOptions};
+use similar::{ChangeTag, TextDiff};
+use std::fs;
+use std::io::Write;
+use std::ops::Add;
+use std::panic::catch_unwind;
+use std::path::{Path, PathBuf};
+use std::time::Instant;
+use tempfile::NamedTempFile;
+
+#[derive(clap::Args)]
+pub(crate) struct Args {
+    /// The root of the repository to reformat
+    pub(crate) dir: PathBuf,
+}
+
+/// The [Jaccard index](https://en.wikipedia.org/wiki/Jaccard_index) can be defined as
+/// ```
+/// J(A, B) = |A∩B| / (|A\B| + |B\A| + |A∩B|)
+/// ```
+/// where in our case `A` is the black formatted input, `B` is the ruff formatted output and the
+/// intersection are the lines in the diff that didn't change. We don't just track intersection and
+/// union so we can make statistics about size changes.
+#[derive(Default, Debug, Copy, Clone)]
+struct JaccardAccumulator {
+    /// The size of `A\B`, the number of lines only in black formatted input
+    black_input: u32,
+    /// The size of `B\A`, the number of lines only in formatted output
+    ruff_output: u32,
+    /// The number of lines unchanged between during formatting
+    intersection: u32,
+}
+
+impl JaccardAccumulator {
+    fn jaccard_index(&self) -> f32 {
+        self.intersection as f32 / (self.black_input + self.ruff_output + self.intersection) as f32
+    }
+}
+
+impl Add<JaccardAccumulator> for JaccardAccumulator {
+    type Output = JaccardAccumulator;
+
+    fn add(self, rhs: JaccardAccumulator) -> Self::Output {
+        JaccardAccumulator {
+            black_input: self.black_input + rhs.black_input,
+            ruff_output: self.ruff_output + rhs.ruff_output,
+            intersection: self.intersection + rhs.intersection,
+        }
+    }
+}
+
+pub(crate) fn main(args: &Args) -> anyhow::Result<()> {
+    let start = Instant::now();
```

Comment:
What's your motivation for creating another command vs integrating it into the `check_formatter_stability`? My main concern is that we start duplicating the code for each command (we probably want to have an option to run the script over multiple projects in the future...)